### ISSUE TITLE
tlog - nbd : multiple tlog client for each vdisk

### ISSUE THREAD

**Issue by iwanbk** [g8os/blockstor#128]
We currently only use one tlog client per vdisk.

Need to check if it is a good idea to make it multiple clients.

the server also need to be adapted to support this.

**Comment by GlenDC** [comment]
Why would we need multiple clients per vdisk? Can't we use a single client for sending blocks in parallel?

**Comment by iwanbk** [comment]
```
Can't we use a single client for sending blocks in parallel?
```
No, if we only have one TCP connection.

**Comment by iwanbk** [comment]
I updated the issue's title and description, our main point is using multiple connections.

**Comment by GlenDC** [comment]
Ah yeah, in that case, this does indeed sound like a good proposition 👍

**Comment by iwanbk** [comment]
my tests using bigger image
- without tlog : 5.8 s

tlog doesn't store the messages -> no flush:
- current code : 11 s
- increase nbd `transactionChCapacity` to 100 (high watermark of tlog flusher) : 10 s
- use 2 tlog clients : 9 s
- use 2 tlog clients, set `transactionChCapacity` to 100: 9 s 

tlog store the messages
- current code : 18 s
- use 2 tlog clients, set `transactionChCapacity` to 100: 18 s -> yes, no improvement

**Comment by iwanbk** [comment]
another interesting test data

with flush, but no capnp encoding. appended all block's data into one big `[]byte`
- current code : 12 seconds
- 2 tlog clients : 11 seconds

**Comment by iwanbk** [comment]
but please note that when i said `2 tlog clients`, the tlog server hasn't been modified to handle the needed changes (i.e. : handle unordered data). This handling itself can slow down the server a bit.

**Comment by iwanbk** [comment]
in latest develop:
- with single client : 12.7 s
- with 2 clients : 11.1 s

i'm still hesitant to use 2 clients now because:
- it added complexity to server (handling unordered data), it will slow the server down by some factor
- not sure about complexity in the client

**Comment by iwanbk** [comment]
i'm just thinking to use two clients:
- one to build the capnp 
- one to send the data to server

lets see how it goes.

**Comment by iwanbk** [comment]
```
i'm still hesitant to use 2 multiple clients now because:
```
but i guess it is something that unavoidable.

**Comment by iwanbk** [comment]
```
i'm just thinking to  add another worker to build the capnp object
```
it only gave small (0.1 s from 12.4 s) improvement.

Instead of using mutiple clients, i think it is better for the client to create multiple (configurable) number of tcp connections. It will be simpler to use.

But before adding multiple tcp connections, we need to add unordered message support to the server.
The C++ version support this behavior by simply using C++ std::map, which is ordered by it's key.

For Go, i can see two alternatives:
- use slice like now, but we sorted the slice.
- use 3rd party data structure : https://github.com/emirpasic/gods, https://github.com/Workiva/go-datastructures, etc....

**Comment by iwanbk** [comment]
depend on #143

**Comment by iwanbk** [comment]
@GlenDC 
trying to answer your question  at
https://github.com/zero-os/0-Disk/pull/209#issuecomment-305071125

> 
> > No, it is single client with multiple TCP connections
> 
> Oh... Well if you can make it work, I'm all happy with that.
> 
> How will that work with the handshake and stuff?
> 
> Can you describe some flows, as I really don't se how that will work.

There won't be any problem with handshaking, we do these:
- each connection do it's own handshaking
- one of the connection failed to handshake, `client.New` returns error

**Comment by iwanbk** [comment]
The first thing we need to do is to encapsulate the socket, reader, and writer into single `connection` struct.
Otherwise the reconnect logic will become a mess.

Marked this as low priority,

**Comment by GlenDC** [comment]
Feel free to move it to milestone alpha-5, if you feel like it isn't required to start using tlog.
